src/BigQuery/QueryResults.php

```diff
@@ -0,0 +1,175 @@
+ */
+class QueryResults
+{
+    /**
+     * @var ConnectionInterface $connection Represents a connection to BigQuery.
+     */
+    private $connection;
+
+    /**
+     * @var The query result's metadata.
+     */
+    private $data;
+
+    /**
+     * @var string The query result's identity.
+     */
+    private $identity;
+
+    /**
+     * @param ConnectionInterface $connection Represents a connection to
+     *        BigQuery.
+     * @param string $jobId The job's ID.
+     * @param string $projectId The project's ID.
+     * @param array $data The query result's metadata.
+     */
+    public function __construct(ConnectionInterface $connection, $jobId, $projectId, array $data)
+    {
+        $this->connection = $connection;
+        $this->data = $data;
+        $this->identity = [
+            'jobId' => $jobId,
+            'projectId' => $projectId
+        ];
+    }
+
+    /**
+     * Retrieves the rows associated with the query and merges them together
+     * with the table's schema. It is recommended to check the completeness of
+     * the query before attempting to access rows.
+     *
+     * Example:
+     * ```
+     * if ($queryResults->isComplete()) {
+     *     $rows = $queryResults->getRows();
+     *     foreach ($rows as $row) {
+     *         echo $row['name'];
+     *     }
+     * }
+     * ```
+     *
+     * @param array $options Configuration options.
+     * @return array
+     */
+    public function getRows(array $options = [])
+    {
+        if (!isset($this->data['rows'])) {
+            return;
+        }
+
+        $schema = $this->data['schema']['fields'];
+
+        while (true) {
+            $options['pageToken'] = isset($this->data['pageToken']) ? $this->data['pageToken'] : null;
+
+            foreach ($this->data['rows'] as $row) {
+                $mergedRow = [];
+
+                foreach ($row['f'] as $key => $value) {
+                    $mergedRow[$schema[$key]['name']] = $value['v'];
+                }
+
+                yield $mergedRow;
+            }
+
+            if (!$options['pageToken']) {
+                return;
+            }
+
+            $this->data = $this->connection->getQueryResults($options + $this->identity);
+        }
+    }
+
+    /**
+     * Checks whether or not the query has been completed. Can be used as a
+     * convenient way to poll for status.
+     *
+     * Example:
+     * ```
+     * // Trigger a network request to check query status
+     * if ($queryResults->isComplete(['force' => true])) {
+     *     return $queryResults->getRows();
+     * }
+     * ```
+     *
+     * @param array $options {
+     *     Configuration options.
+     *
+     *     @type bool $force If true fetches fresh data, otherwise returns data
+     *           stored locally if it exists.
+     * }
+     * @return bool
+     */
+    public function isComplete(array $options = [])
```

Comment:
You are absolutely right. I was worried about having to make multiple network requests because a `jobs.get` returns a full job resource whereas a `jobs.getQueryResults` returns the rows we need - but at this point we would already have the initial set of rows anyways. Good call :).